pymc_marketing/clv/models/pareto_nbd.py

```diff
@@ -0,0 +1,537 @@
+        if frequency is None:
+            frequency = self._frequency
+        if recency is None:
+            recency = self._recency
+        if T is None:
+            T = self._T
+
+        return to_xarray(customer_id, frequency, recency, T)
+
+    @staticmethod
+    def _logp(
+        r: np.ndarray,
+        alpha: np.ndarray,
+        s: np.ndarray,
+        beta: np.ndarray,
+        x: xarray.DataArray,
+        t_x: xarray.DataArray,
+        T: xarray.DataArray,
+    ) -> xarray.DataArray:
+        """
+        Utility function for using ParetoNBD log-likelihood in predictive methods.
+        """
+        # Add one dummy dimension to the right of the scalar parameters, so they broadcast with the `T` vector
+        pareto_dist = ParetoNBD.dist(
+            r=r.values[..., None],
+            alpha=alpha.values[..., None],
+            s=s.values[..., None],
+            beta=beta.values[..., None],
+            T=T.values,
+        )
+        values = np.vstack((t_x.values, x.values)).T
+        loglike = pm.logp(pareto_dist, values).eval()
+        return xarray.DataArray(data=loglike, dims=("chain", "draw", "customer_id"))
+
+    def expected_purchases(
+        self,
+        future_t: Union[float, np.ndarray, pd.Series],
+        customer_id: Union[np.ndarray, pd.Series] = None,
+        recency: Union[np.ndarray, pd.Series] = None,
+        frequency: Union[np.ndarray, pd.Series] = None,
+        T: Union[np.ndarray, pd.Series] = None,
+    ) -> xarray.DataArray:
+        r"""
+            Given :math:`recency`, :math:`frequency`, and :math:`T` for an individual customer, this method returns the
+            expected number of future purchases across :math:`future_t` time periods.
+
+            If no customer data is provided, probabilities for all customers in model fit dataset are returned.
+
+            Calculate the expected number of repeat purchases up to time t for a
+            randomly choose individual from the population, given they have
+            purchase history (frequency, recency, T).
+
+            See equation (41) from [2]_.
+
+            Adapted from lifetimes package
+        https://github.com/CamDavidsonPilon/lifetimes/blob/41e394923ad72b17b5da93e88cfabab43f51abe2/lifetimes/fitters/pareto_nbd_fitter.py#L242
+            Parameters
+            ----------
+            future_t: array_like
+                times to calculate the expectation for.
+            customer_id: array_like
+                Customer labels.
+            recency: array_like
+                Number of time periods between the customer's first and most recent purchases.
+            frequency: array_like
+                Number of repeat purchases per customer.
+            T: array_like
+                Number of time periods since the customer's first purchase.
+                Model assumptions require T >= recency.
+        """
+
+        x, t_x, T = self._process_customers(customer_id, frequency, recency, T)
```

Comment:
Can the `customer_id` variable be `None` (as defined in the signature)? Because this is incompatible with the types passed to `self._process_customers` where `None` is not explicitly allowed.